### DISCUSSION TITLE
Adding a reverse function of `avro.parse()`

### DISCUSSION THREAD

**Discussion by TanimSk** [hitblast/avro.py#4] [General]
**Details**:
Add a reverse parser, for example:

```output = avro.reverse_parse("ভালো")```
```print(output)```

Desired output: `valo`

**Reason**:
Actually I am building a chatbot for processing Bangla language typed in English alphabets. Cause people nowadays chat in bangla using English alphabets for saving time and efforts. Moreover, adding this functionality will give this project more flexibility.

I highly appreciate that you guys are still maintaining this project.

**Comment by hitblast** [comment]
Hi there! This seems like a pretty legit idea to work on, so definitely will look into it. Also, if you have any thoughts about making this feature possible then don't hesitate to note it down here :D The code's open so hope that you will also have a look at it to brainstorm a little bit.

Last but not least, <3

**Comment by TanimSk** [comment]
glad to have a wave back!
I will surely look into the codes 🥂

**Comment by hitblast** [comment]
Alright 👊 If you're interested, then we can convert this Issue into a GitHub Discussion and then develop the feature together, since it'll definitely require some reverse engineering over the already-existing Avro Dictionary.

**Comment by baseplate-admin** [comment]
>  I will surely look into the codes 🥂

lemme know if you run into errors.

The approach i mentioned should work :D 

@TanimSk if you want to we can work together on making this a better project ...

**Comment by baseplate-admin** [comment]
Looking over the codes I have a few questions

@hitblast,

Is it possible not to bundle dependencies such as `click` and `rich` and `pyclip` ?

Say i want to bundle it inside a django project. I dont have a hard dependency on either `rich` or `click`. Adding a CLI toolkit into a parser seems a bit useless no?

Suggestions :
* Release a separate package named maybe `avro.py-cli`
* Freeze the core so that it works with all the python versions


----

Also it seems that you are not using type-hints properly.. I have seen many cases of Any ( which can be strongly typed )

**Reply by hitblast** [reply]
@baseplate-admin 

Speaking of the first point that you have made, the idea seems fair enough. One viable option could be to implement the idea using the "optional features" mechanism of setuptools.
```bash
$ pip install avro.py[cli]
```
Or, as you said, we could make a separate package that wraps around avro.py to implement terminal functionality along with some nifty clipboard shortcuts.

Oh about the second point, the only Python 3.10 features used in this library are, for now, **union operators** and a couple syntactical changes. Backwards compatibility can easily be added to it by simply *not* using those features.

The final one, yeah I was in a hurry while fixing some of the obsolete code from [kaustavdm's pyAvroPhonetic](https://github.com/kaustavdm/pyAvroPhonetic) so I didn't get proper time to utilize on fixing the type hinting. `Any` seemed a quick option.

**Reply by baseplate-admin** [reply]
> Speaking of the first point that you have made, the idea seems fair enough. One viable option could be to implement the idea using the "optional features" mechanism of setuptools.
 
 I am +1 on this.



Now we can do something like 
* publish `avro.py-cli`
* `pip install avro.py-cli` # Which will indirectly be installed with `avro.py[cli]`

**Reply by hitblast** [reply]
https://github.com/hitblast/avro.py-cli

Already done (almost) packaging the extras into a separate package. Proceeding with the optional features mechanism could initially break avro.py itself because of unmet dependencies within the unused scripts.

**Reply by baseplate-admin** [reply]
Woooh you are fast :D  ( Congratulations | This satisfies my and @TanimSk need perfectly | Thanks :D  )

**Reply by hitblast** [reply]
> Woooh you are fast :D ( Congratulations | This satisfies my and @TanimSk need perfectly | Thanks :D )

Glad to help! Now what that remains is the making of `avro.reverse()` (or whatever) and I will ship the latest version with no unnecessary dependencies :D

**Reply by TanimSk** [reply]
@hitblast i've gone thru your codes. swapping `find` with `replace` without rules kinda works
https://github.com/hitblast/avro.py/blob/6188ff95256336d531deee394d014cec4ab2e66e/avro/main.py#L177-L181
i guess we need to rewrite the `rules` of avro dictionary for that reverse function to gain more accuracy

**Reply by hitblast** [reply]
Yeah that's an option, how about we write separate rules for the reversal?

**Reply by TanimSk** [reply]
I was thinking about that too. But how should we get into it? Did they used any tools for generating the dict or should we make one?
Cause editing manually will be a pain in the butt.

**Reply by hitblast** [reply]
Well they surely made it from scratch. It'd be fun if we could reach out to the creators but they might just be busy. Manual is the way to go, for now. How's the accuracy on the method you proposed?

**Reply by TanimSk** [reply]
The accuracy is not so good. For example:
```python
print(reversed.parse_reverse("বক"))
```
`output: bk`
`expected output: bok`

we need to reconstruct the rules for this function. 
For instance, on the case of  `বক`, if `ক` is `not preffix` then append `o` before `k`. So it gets `বক` -> `bok`
And, if the `ক` is `preffix` then simply append `k`. Like, `কলম` -> `kolom`

**Reply by hitblast** [reply]
Seems reasonable. We'll just have to play with the prefixes so that it ends up being understandable to us and not necessarily Avro itself. How do you plan to deploy your changes to the code?

**Reply by TanimSk** [reply]
Maybe open a development branch in your repo? Then gimme access to it. 
Or I will fork it and send PR to your repo?

**Reply by hitblast** [reply]
PR might be preferable in this case because that'd indicate your contribution to this package :D

**Reply by baseplate-admin** [reply]
@TanimSk 

Can you add tests with your code? 

Also another recommendation

Use `reverse_parse()` as your function name

---


> PR might be preferable in this case because that'd indicate your contribution to this package :D

If he commits directly he can also get his contribution badge

**Reply by TanimSk** [reply]
@baseplate-admin  I will of course add tests. But  for now, I wanna focus on the new avro dict construction. I need some time to figure it out.

**Reply by hitblast** [reply]
@TanimSk Okay so if everything's sorted out, I hope that you won't have much problem developing the new dictionary. Make sure to not override the key:value pairs used for the original `parse()` function or that could potentially break how avro can parse same texts even with multiple inputs (bk -> বক but bok is also বক). Best of luck 👍 and if you need any assistance, ping me.

> If he commits directly he can also get his contribution badge

@baseplate-admin A PR might just be more visible in his activity feed.

**Reply by TanimSk** [reply]
@hitblast got it. I am working on separate files, so, don't worry about breaking up things ✌️

**Reply by hitblast** [reply]
@TanimSk Alright! Looks like you're pretty cheered up there. Make sure to sync your fork before you continue to develop on your feature request.

**Comment by hitblast** [comment]
@TanimSk Let us know about your progress once you're free :P

**Comment by TanimSk** [comment]
@hitblast Actually I have bunch of projects in this month. I'll work on avro as soon as I get free, btw do you use any social media? I would love to connect with you. You can mail me if you dont wanna make your account public :D

**Reply by hitblast** [reply]
Ah being public is something I won't hate though. My mail address is given in my GitHub bio, and other social links include my Discord account. You could apparently use Discord to contact me if you're familiar!

**Comment by hitblast** [comment]
@baseplate-admin @TanimSk Well, I have launched **avro.py 0.1.5** ahead of the release of the feature Tanim's currently working on in order to force compatibility with avro.py-cli. I hope we'll have a version 0.1.6 as soon as he's done prototyping and we're done doing fault-checks!

**Reply by TanimSk** [reply]
@hitblast hey, hitblast, hopefully, you're doing great. 
Just to let you know about the progress of avro reversed function .
```python
from avro import reversed

print(reversed.reverse_parse('আমার সোনার বাংলা আমি তোমায় ভালোবাসি'))
print(reversed.reverse_parse('পাখি পাকা পেঁপে খায়'))
print(reversed.reverse_parse('তোমার বাসায় সবাই কেমন আছে?'))
print(reversed.reverse_parse('তুমি কি খেতে ভালোবাসো ?'))
print(reversed.reverse_parse('আমি যা বলি, তাই করি'))
print(reversed.reverse_parse('আসসালামু আলাইকুম ভাইজান । কি খবর ?'))
print(reversed.reverse_parse("ঘুঘু"))

print(reversed.reverse_parse('স্বপ্ন মাত্রই সত্যি হয়'))
print(reversed.reverse_parse('ফেইসবুক'))
print(reversed.reverse_parse('চৈত্রী এসে দাঁড়িয়েছে সন্ধ্যা থেকে কবিতার চোখের ভিতরে দারুণ উজ্জ্বল হয়ে'))
```

output :
```
amar sonar bangla ami tomay bhalobasi
pakhi paka pepe khay
tomar basay sbai kemn ase?
tumi ki khete bhalobaso ?
ami za bli, tai kri
assalamu alaikum bhaijan . ki khbr ?
ghughu
sbpn matri stzi hz়
pheisbuk
choitri ese dariyese sndhza theke kbitar chokher bhitre darun ujjbl hye
```
so, it works just fine with regular sentences. But it gets messy when the words and sentences get complex.

**Reply by hitblast** [reply]
How about parsing the already-reversed text and see if the output matches the previous input? If it does then we're obviously close. Also, I would love to have a look at the dictionary structure :D

**Reply by TanimSk** [reply]
Yea see my forked repo.
I did tried parsing the already reversed text, the outputs were not readable. Look at my codes you will get it.

**Reply by baseplate-admin** [reply]
I have another suggestion.


Rename function to `parse`

```python
from avro import reverse

reverse.parse()
```

Or

```python
from avro import parse

parse.reverse()

```

----

And about the dictionary problem


We can do this

```python
{
"english": " Facebook",
"bangla" : "ফেসবুক"
}
```

Pros : 
Both ridmik and avro use this method

Cons : 
We need to hardcode these values.

**Reply by hitblast** [reply]
Saw the repository just now. Have a few thoughts.

How about instead of making a completely separate dictionary for the reversed text, we add a new 'reversed' key:value inside `avro/utils/avrodict.json`? 

```json
{
    "find": "kT",
    "replace": "ক্ট",
    "reversed": "kt"
}
```

- The new flag will contain the proper, readable, plain English version of the 'find' flag. If necessary, the flag must be added to a select number of characters.

- If there's no such flag for a character, the code will proceed to use the default 'find' key for replacing the desired Bangla characters.

- In this case, we don't really need to reverse like we need to parse again, we're focusing on readability so this point is not worthwhile.

Oh also, the syntax for both types of methods could be combined into one single, `parse` module.
```python
from avro import parse

bangla = parse.do('bulbuli')  # বুলবুলি
reversed = parse.reverse(bangla)  # bulbuli
```

Obviously loved your approach in this case though <3

**Reply by TanimSk** [reply]
@hitblast just saw your edited msg. It makes much more sense. I will rework on that.
@baseplate-admin hardcoding all words like that will make the query procedure slower, do you have any other approach?
Cause you see if I use `কলম` it will loop through three times `ক` `ল` `ম` thru the dictionary which has more than 1200+ key mapping

**Reply by baseplate-admin** [reply]
> ```python
> from avro import parse
> bangla = parse.do('bulbuli')  # বুলবুলি
> reversed = parse.reverse(bangla)  # bulbuli
> ```

I am -1 on this.

 We can do something like ( this is reallly hard to implement properly ) 

```python
parse('bulbuli').reverse() # বুলবুলি
reversed('বুলবুলি') # bulbuli
```

The same api list uses. Which makes the code more pythonic imho


Or ( this is better path ) 

```python
from avro import parser

parser.parse_english('bulbuli')

parser.parse_bangla('বুলবুলি')
```
---
@TanimSk wait no you are wrong.

Approach here is to use

```python
dict = {"facebook": "ফেসবুক"}

dict.get('facebook')
```

if you use dict.get(key) its not looping actually.  Its doing a hashmap lookup ( which is why the entire python interpreter is using dictionary for ast ) 

https://stackoverflow.com/questions/8023306/get-key-by-value-in-dictionary

Heres how to do exact thing in reverse

**Reply by TanimSk** [reply]
okay, so we can do put a key named `exceptions`
```
"data": {
        "patterns": [
            {
                "find": "bhl",
                "replace": "ভ্ল"
                "reversed": "bhl"
            },
            ...
        ]
        "exceptions":
            {
                "ফেসবুক": "facebook",
                ...
            }
}
```
so, the workflow will be,
```python
word = dict.get("ফেসবুক", None)
if not word == None:
   new_word =  iterate_through_dict()
   return new_word
else:
   return word
```

**Reply by baseplate-admin** [reply]
> okay, so we can do put a key named exceptions

Yes but it should not be in the same JSON. We should split out json file. 


In fact i am -1 on using JSON.


Lets just use a python dictionary. Its faster and doesnt need to tokenize stuff

> ```python
> word = dict.get("ফেসবুক")
> if not word == None:
>   new_word =  iterate_through_dict()
>   return new_word
> else:
>   return word
> ```

This is better

```python
word = dict.get("ফেসবুক")
if not word:
   word = iterate_through_dict()
return word
```

**Reply by TanimSk** [reply]
umm actually we ARE converting JSON to python dict, not using JSON
https://github.com/hitblast/avro.py/blob/908c29bee68b7a1a5482f7395bff1330daf527a6/avro/config.py#L49-L51
```python
type(AVRO_DICT)
```
returns `<class 'dict'>`

**Reply by baseplate-admin** [reply]
> umm actually we ARE converting JSON to python dict, not using JSON

Thats what we shouldn’t do. If you do inspect on JSON module you will see how many calls it make under the hood.
Why not just use a python dictionary? Whats the benefit JSON is giving us ( JSON is not even a python standard | Marshall is ) 

Also loading json and converting it to dictionary on every call seems like a very inefficient design choice no?

**Reply by hitblast** [reply]
@TanimSk @baseplate-admin 

- Based on both of your replies, we could hardcode the dictionary into a separate `avrodict.py` file. This will make the dictionary native to Python's interpreter, increasing performance while adding syntax highlighting to distinguish errors.

- I've come upon a much cleaner approach with the syntax:
```python
import avro

some_random_english = 'bulbuli'

# Eng to Ban
banglified = avro.parse(some_random_english)

# Ban to Eng
reversed = avro.reverse(banglified)

# Singular imports make up for a cleaner code style :D
```

- I like the 'exceptions' idea, except I think that it should be added based on user's usage data. If you have used an Avro application (iAvro / Avro on Windows) before then you've most probably noticed that Avro saves a copy of the word you chose from Avro's suggestions list, then corrects them when you type them afterwards.

**Comment by hitblast** [comment]
@TanimSk Based on our previous conversation, should I start working on the "Pythonized" Avro Dictionary? This way you could straight up start working on the newer suggested patterns.

**Reply by TanimSk** [reply]
Yea sure @hitblast, lmk when you're done

**Reply by hitblast** [reply]
Alright, thanks for the confirmation from your side.

**Comment by TanimSk** [comment]
Hi there, @hitblast. Hopefully, all is going well.
I've updated my repo according to the basis of your and @baseplate-admin  suggestions. Also have added some rules to enhance the output result :D
The current dictionary structure looks like this:
```python
{
       ...
        [
             {
                "find": "A`",
                "replace": "া",
                "reverse": "a"
            },
            {
                "replace": "আ",
                "reverse": "a"
            },
            ....
       ],

        "exceptions":{
            "ফেসবুক": "Facebook",
            "গুগল": "Google",

        },

        "vowel": "aeiou",
        "consonant": "bcdfghjklmnpqrstvwxyz",
        "casesensitive": "oiudgjnrstyz",
        "number": "0123456789",
        
        # For reverse parsing
        "shorborno": "অআইঈউঊঊএঐওঔ",
        "kar": {'া', 'ি', 'ী', 'ু', 'ূ', 'ৃ', 'ে', 'ৈ', 'ো', 'ৌ'},
        "ignore": {"ঁ", "।", "?", ".", "-", ";"}
}

```

Some examples:
```python
import avro
print(avro.reverse('তোমার বাসায় সবাই কেমন আছে?'))
print(avro.reverse('আমি যা বলি তাই করি'))
print(avro.reverse('আসসালামু আলাইকুম ভাইজান। কি খবর?'))
print(avro.reverse('তুমি কি খেতে ভালোবাসো?'))
print(avro.reverse("কাকতাড়ুয়া"))
print(avro.reverse("কাঁকন"))
print(avro.reverse("রাহিম, তোমাকে করিম ডাকছে। এখন কি রওনা দেবে?"))
print(avro.reverse('আমার সোনার বাংলা।'))
print(avro.reverse('গুগল'))
print(avro.reverse('ফেসবুক'))
```
output:
```
tomar basay sobai kemon ase?
ami za boli tai kori
assalamu alaikum bhaijan. ki khobor?
tumi ki khete bhalobaso?
kaktaruya
kakon
rahim, tomake korim dakse. ekhon ki rowna debe?
amar sonar bangla.
Google
Facebook
```
Hopefully, you'll sync up soon and have a look at my repo. If everything seems perfect, I will send a PR :clinking_glasses:

**Reply by baseplate-admin** [reply]
Bit of an advice. 

Open up the dictionary 

```python
D = {} # closed dictionary 
```
```python
D = {  # open dictionary
      }
```

Helps improve readability.

**Reply by hitblast** [reply]
@baseplate-admin Like this?
```python
D = {
  "a": 1,
  "b": 2
}
```
Yeah indeed it does improve readability by a decent margin.

@TanimSk I guess you can take this as a suggestion from our side. Everything else looks pretty neat :D

**Reply by baseplate-admin** [reply]
@hitblast yep you are right :).


@tanimsk use a tool like black and add a trailing comma. Black should do the rest for you

**Reply by hitblast** [reply]
@baseplate-admin Not sure if we wanna add black cause avro.py already uses flake8.

**Reply by baseplate-admin** [reply]
Black is a text formatter.

Flake8 is a linter. 

Both serve a different purpose :)

**Reply by hitblast** [reply]
@baseplate-admin Was just saying if they could interrupt each other :P

**Reply by baseplate-admin** [reply]
Take a look at this [project](https://github.com/baseplate-admin/CoreProject/blob/3916910dbb6d53b0570067bbb03b333bc2446825/backend/). I use them together and they dont seem to be conflicting :)


Just a small suggestion to improve quality.

( but theres a tool called isort that conflicts with black | for which you need to cofigure a switch to support black ) 
( in fact black is the defacto way to format files )

**Reply by TanimSk** [reply]
@baseplate-admin idk how much familiar you are with black, 
I get this result after formatting 😕 however, I opened the dictionary braces, sending a PR
![image](https://user-images.githubusercontent.com/67966003/194771202-784482b4-77a4-4724-9fbc-10e470034f2d.png)

**Reply by hitblast** [reply]
Doesn't look bad I guess.

**Reply by baseplate-admin** [reply]
@TanimSk how much are you familiar with formatting 🙁
 
Btw all of my projects and every piece of code i wrote uses black :/ 

---

Btw @hitblast look into https://pre-commit.ci 

It automates some of the hassle that @TanimSk might need to do manually

**Reply by hitblast** [reply]
@baseplate-admin It's actually possible with GitHub Actions as well. I just gotta let the workflow have permissions to write content and remove the `--check` flag from the command.

**Reply by baseplate-admin** [reply]
While its possible to use github actions. Your minutes will quickly run out when you have a large number of PR/ a large number of repositories... 

Pre-Commit CI offers unlimited minutes for any open source project :)

**Reply by hitblast** [reply]
@baseplate-admin Hello! Sorry I got busy during exams.

GitHub Actions also has unlimited runtime for public repositories. The limit kicks in at private repositories only.

**Reply by baseplate-admin** [reply]
Hi there,

I also got busy due to exam .

Now back to topic at hand.

I recommended https://pre-commit.ci for 3 reasons.
1. [Its a lot faster than `GitHub Actions`.](https://pre-commit.ci/)
2. Its interoperable with [`Github Actions`](https://github.com/pre-commit/action). 
3. Its cross language and open source :D 

You dont need to completely go in one route or another. Going pre-commit means you can not only get a code checker but also you can run those using github actions.

**Reply by hitblast** [reply]
@baseplate-admin Alright, I will look into it.

**Comment by hitblast** [comment]
Well it's been long since the feature has been implemented and released, so I guess I'll close this discussion here.
